File: arduino/serial_interface.cpp
#include <Arduino.h>

#include "serial_interface.h"

SerialInterface::SerialInterface(int serial_rate, unsigned long timeout) {
  this->timeout = timeout;
  Serial.begin(serial_rate);
  Serial.setTimeout(this->timeout);
}

bool SerialInterface::IsConnected() {
  if (Serial) {
    return true;
  }
  else {
    return false;
  }
}

void SerialInterface::SetRecvTimout(unsigned long timeout) {
  this->timeout = timeout;
  Serial.setTimeout(this->timeout);
}

int64_t SerialInterface::Available() {
  return Serial.available();
}

bool SerialInterface::WaitForBytes(size_t num_bytes, unsigned long timeout) {
  unsigned long startTime = millis();
  //Wait for incoming bytes or exit if timeout
  while ((Serial.available() < num_bytes) && (millis() - startTime < timeout)){}
}

void SerialInterface::ReadBuffer(byte* buffer, size_t size) {
  this->WaitForBytes(size, this->timeout);
  byte data = 0;
  for (size_t i = 0; i < size; i++) {
    data = Serial.read();   
    buffer[i] = data;
  }
}

int8_t SerialInterface::Read_i8() {
  byte buffer[1];
  this->WaitForBytes(1, this->timeout);
  this->ReadBuffer(buffer, 1);
  return (int8_t) buffer[0];
}

int16_t SerialInterface::Read_i16() {
  byte buffer[2];
  this->WaitForBytes(2, this->timeout);
  this->ReadBuffer(buffer, 2);
  return (((int16_t) buffer[0]) << 0 & 0x00ff) | 
         (((int16_t) buffer[1]) << 8 & 0xff00);
}

int32_t SerialInterface::Read_i32() {
  byte buffer[4];
  this->WaitForBytes(4, this->timeout);
  this->ReadBuffer(buffer, 4);
  return (((int32_t) buffer[0]) << 0  & 0x000000ff) | 
         (((int32_t) buffer[1]) << 8  & 0x0000ff00) | 
         (((int32_t) buffer[2]) << 16 & 0x00ff0000) | 
         (((int32_t) buffer[3]) << 24 & 0xff000000);
}

void SerialInterface::WriteBuffer(byte* buffer, size_t size) {
  Serial.write(buffer, size);
}

void SerialInterface::Write_i8(int8_t data) {
  byte buffer[1] = {
    (byte) data
  };
  this->WriteBuffer(buffer, 1);
}

void SerialInterface::Write_i16(int16_t data) {
  byte buffer[2] = {
    (byte) ((data >> 0) & 0x00ff),
    (byte) ((data >> 8) & 0x00ff)
  };
  this->WriteBuffer(buffer, 2);
}

void SerialInterface::Write_i32(int32_t data) {
  byte buffer[4] = {
    (byte) ((data >> 0)  & 0x000000ff),
    (byte) ((data >> 8)  & 0x000000ff),
    (byte) ((data >> 16) & 0x000000ff),
    (byte) ((data >> 24) & 0x000000ff)
  };
  this->WriteBuffer(buffer, 4);
}

void SerialInterface::println(const char* str) {
  Serial.println(str);
  this->Write_i8(0);
}

void SerialInterface::print(const char* str) {
  Serial.print(str);
  this->Write_i8(0);
}

File: cli/eeprom_programmer/serial_interface.py
import struct
import serial
import time
import glob
import sys


class SerialInterface(object):
    _BIT_ORDER = "<"

    def __init__(self, port, baudrate, timeout, write_timeout, debug=False):
        self._port = port
        self._baudrate = baudrate
        self._timeout = timeout
        self._write_timeout = write_timeout
        self._debug = debug
        self._serial = None

    def __enter__(self):
        self.open()
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        try:
            self.close()
        except Exception as e:
            # Raise exception only if we won't shadow any previous one.
            if exc_type is None and exc_val is None and exc_tb is None:
                raise e

    def open(self):
        if self._serial is None:
            self._serial = serial.Serial(
                port=self._port,
                baudrate=self._baudrate,
                timeout=self._timeout,
                writeTimeout=self._write_timeout
            )

            # The serial port might take a while before we can send/recv any data
            time.sleep(2)

    def close(self):
        if self._serial is not None:
            self._serial.close()
            self._serial = None

    def read_buffer(self, length: int) -> bytes:
        buffer = self._serial.read(length)
        if self._debug:
            print("<", buffer)

        return buffer

    def read_i8(self) -> int:
        buffer = self.read_buffer(1)
        return struct.unpack(self._BIT_ORDER + "B", buffer)[0]

    def read_i16(self) -> int:
        buffer = self.read_buffer(2)
        return struct.unpack(self._BIT_ORDER + "H", buffer)[0]

    def read_i32(self) -> int:
        buffer = self.read_buffer(4)
        return struct.unpack(self._BIT_ORDER + "L", buffer)[0]

    def read_string(self):
        buffer = []
        byte = b""
        while byte != b"\x00":
            byte = self.read_buffer(1)
            buffer.append(byte)

        return str(b"".join(buffer), encoding="ASCII")

    def write_buffer(self, buffer: bytes) -> None:
        if self._debug:
            print(">", buffer)

        self._serial.write(buffer)

    def write_i8(self, value: int):
        buffer = struct.pack(self._BIT_ORDER + "B", value)
        self.write_buffer(buffer)

    def write_i16(self, value: int):
        buffer = struct.pack(self._BIT_ORDER + "H", value)
        self.write_buffer(buffer)

    def write_i32(self, value: int):
        buffer = struct.pack(self._BIT_ORDER + "L", value)
        self.write_buffer(buffer)

    def write_string(self, string: str) -> None:
        buffer = bytes(string, encoding="ASCII")
        buffer = buffer + b"\x00"
        self.write_buffer(buffer)


# From https://stackoverflow.com/questions/12090503/listing-available-com-ports-with-python
def get_serial_ports():
    """
    Lists serial ports.
    :return: ([str]) A list of available serial ports
    """
    if sys.platform.startswith('win'):
        ports = ['COM%s' % (i + 1) for i in range(256)]
    elif sys.platform.startswith('linux') or sys.platform.startswith('cygwin'):
        # this excludes your current terminal "/dev/tty"
        ports = glob.glob('/dev/tty[A-Za-z]*')
    elif sys.platform.startswith('darwin'):
        ports = glob.glob('/dev/tty.*')
    else:
        raise EnvironmentError('Unsupported platform')

    results = []
    for port in ports:
        try:
            s = serial.Serial(port)
            s.close()
            results.append(port)
        except (OSError, serial.SerialException):
            pass
    return results
